libraries/AP_Compass/CompassCalibrator.cpp

```diff
@@ -696,3 +699,140 @@ void CompassCalibrator::CompassSample::set(const Vector3f &in) {
+    v.rotate_inverse(_orientation);
+
+    // rotate to body frame for this rotation
+    v.rotate(r);
+
+    Vector3f rot_offsets = _params.offset;
+    rot_offsets.rotate_inverse(_orientation);
+
+    rot_offsets.rotate(r);
+    
+    v += rot_offsets;
+
+    Matrix3f rot = sample.att.get_rotmat();
+
+    Vector3f efield = rot * v;
+    return efield;
+}
+
+/*
+  calculate compass orientation using the attitude estimate associated with each sample
+ */
+bool CompassCalibrator::calculate_orientation(void)
+{
+    if (!_auto_orientation) {
+        // only calculate orientation for external compasses
+        return true;
+    }
+
+    float sum_error[ROTATION_MAX] {};
+
+    for (enum Rotation r = ROTATION_NONE; r<ROTATION_MAX; r = (enum Rotation)(r+1)) {
+        // calculate the average implied earth field across all samples
+        Vector3f total_ef {};
+        for (uint32_t i=0; i<_samples_collected; i++) {
+            Vector3f efield = calculate_earth_field(_sample_buffer[i], r);
+            total_ef += efield;
+        }
+        Vector3f avg_efield = total_ef / _samples_collected;
+
+        // now calculate the square error for this rotation against the average earth field
+        for (uint32_t i=0; i<_samples_collected; i++) {
+            Vector3f efield = calculate_earth_field(_sample_buffer[i], r);
+            float err = (efield - avg_efield).length_squared();
+            sum_error[r] += err;
+        }
+    }
+
+    // find the rotation with the lowest square error
+    enum Rotation besti = ROTATION_NONE;
+    float bestv = sum_error[0];
+    for (enum Rotation r = ROTATION_NONE; r<ROTATION_MAX; r = (enum Rotation)(r+1)) {
+        if (sum_error[r] < bestv) {
+            bestv = sum_error[r];
+            besti = r;
+        }
+    }
+
+    // consider this a pass if the best orientation is 4x better
+    // square error than 2nd best
+    float second_best = besti==ROTATION_NONE?sum_error[1]:sum_error[0];
+    for (enum Rotation r = ROTATION_NONE; r<ROTATION_MAX; r = (enum Rotation)(r+1)) {
+        if (r != besti) {
+            if (sum_error[r] < second_best) {
+                second_best = sum_error[r];
+            }
+        }
+    }
+    bool pass;
+    if (besti == _orientation) {
+        // if the orientation matched then allow for a low threshold
+        pass = true;
+    } else {
+        pass = (second_best / bestv) > 4;
+    }
+    if (!pass) {
+        gcs().send_text(MAV_SEVERITY_CRITICAL, "Bad orientation: %u %.1f\n", besti, second_best/bestv);
+    } else if (besti == _orientation) {
+        // no orientation change
+        gcs().send_text(MAV_SEVERITY_INFO, "Good orientation: %u %.1f\n", besti, second_best/bestv);
+    } else {
+        gcs().send_text(MAV_SEVERITY_INFO, "New orientation: %u was %u %.1f\n", besti, _orientation, second_best/bestv);
+    }
+
+    if (!pass) {
+        // give an indication of orientation failure by showing a very high fitness
```

Comment:
This is going to lead to a lot of user/GCS confusion. We should be handling this with a proper MAVLink extension instead.